nemo_aligner/algorithms/critic_server_trainer.py

```diff
@@ -59,22 +63,32 @@ def __init__(self, cfg, model, optimizer, scheduler, logger, ckpt_callback, gbs)
         self.gbs = gbs
         self.step = 0
 
+        self.tokenize_func = tokenize_func
+
         # server parameters
         self.combine_rm_and_critic_server = cfg.combine_rm_and_critic_server
+        self.max_queue_delay_microseconds = cfg.get("max_queue_delay_microseconds", 2000)
+        self.strip_sequence_length_to_multiple = cfg.get("strip_sequence_length_to_multiple", None)
+
+        inference_micro_batch_size = cfg.inference_micro_batch_size
+        if isinstance(inference_micro_batch_size, int):  # for backward compatibility
+            inference_micro_batch_size = [inference_micro_batch_size]
+        self.preferred_batch_sizes = [
+            item * parallel_state.get_data_parallel_world_size() for item in inference_micro_batch_size
+        ]
+        self.pad_batch_to_multiple = min(self.preferred_batch_sizes)
```

Comment:
Why aren't we using the model's `forward_micro_batch_size` for `pad_batch_to_multiple`?

Reply:
it is possible for `inference_micro_batch_size` to be smaller than `forward_micro_batch_size` and in that case we lower the forward micro batch size to as low as possible using this logic: https://github.com/NVIDIA/NeMo-Aligner/blob/geshen/critic_speedup/nemo_aligner/models/nlp/gpt/megatron_gpt_reward_model.py#L380. So I want to pad it to as little as user specified.

this is to prevent the case where pytrition batches a number that is smaller than `forward_micro_batch_size`, but there is an argument to be made that we should pad it.

Reply:
Is there actually a use case for `inference_micro_batch_size < forward_micro_batch_size`? (shouldn't one just use a lower `forward_micro_batch_size` if they want small inference batches?)

My concern with using `min(self.preferred_batch_sizes)` is that if this is set to a high value (which seems to be possible due to the PP case mentioned in the config comments), we may end up doing a lot of useless padding when the batch is actually small.

That being said, after giving it more thought I actually wonder if what we want is simply using `self.pad_batch_to_multiple = parallel_state.get_data_parallel_world_size()`? Why do we care about trying to match `inference_micro_batch_size` if we have less data to process?

Reply:
> Is there actually a use case for inference_micro_batch_size < forward_micro_batch_size? (shouldn't one just use a lower forward_micro_batch_size if they want small inference batches?)

I was imagining the case where we want to allow small batches to flow through to prevent the case of pytriton waiting a long time to get a bigger batch we'd want and then we have to eat all the padding costs after it doesn't succeed. that being said I think the perf gain for this should be extremely minor, and this really complicates all our logic.

Given your suggestion below of 
```
If inference_batch_size < forward_micro_batch_size => use a single batch of size inference_batch_size
If inference_batch_size > forward_micro_batch_size => pad to multiple of forward_micro_batch_size
```
I think we should do this and just set pad batch to multiple to the DP size. I think it's much cleaner that way.

I fixed the pad_batch_to_multiple logic here: 39df6591f0ceae02a92c3d51aa6f279df96a7fab

Reply:
Agreed!